fedimint-core/src/task/jit.rs

```diff
@@ -0,0 +1,208 @@
+use std::sync::Arc;
+
+use fedimint_logging::LOG_TASK;
+use futures::Future;
+use tokio::select;
+use tokio::sync::{self, oneshot};
+use tracing::trace;
+
+use super::MaybeSend;
+
+/// A value that initializes eagerly in parallel
+#[derive(Debug, Clone)]
+pub struct Jit<T> {
+    // on last drop it lets the inner task to know it should stop, because we don't care anymore
+    _cancel_tx: sync::mpsc::Sender<()>,
+    // since joinhandles, are not portable, we use this to wait for the value
+    // Note: the `std::sync::Mutex` is used intentionally, since contention never actually happens
+    // and we want to avoid cancelation because of it
+    val_rx: Arc<std::sync::Mutex<Option<tokio::sync::oneshot::Receiver<T>>>>,
+    val: sync::OnceCell<T>,
+}
```

Comment:
That's a lot of complexity for something that *seems* fairly simple to implement with a [`watch`](https://docs.rs/tokio/latest/tokio/sync/watch/index.html) channel. My approach would be:

```rust
struct Jit<T> {
    val: tokio::sync::watch::Receiver<Option<T>>
}

impl<T> Jit<T> {
    fn new(f: impl Future<T>) {
        let (sender, receiver) = tokio::sync::watch::channel();
        spawn(async move {
            select!{
                val = f => {
                    // TODO: handle error
                    sender.send(Some(val));
                },
                _ = sender.closed() => {
                   // Value no longer needed
                },
            };
        });

        Jit {
            val: receiver,
        }
    }

    async fn get(&self) -> T {
        self.val
            .wait_for(Option::is_some)
            .await
            .expect("Will only return once value is set")
    }
}
```

Maybe I'm missing some functionality?

Reply:
> I'd like to add `Into::into::<T>` so I can convert the `Jit<T>` to `T` eventually

I think that's a bad idea since it mixes sync+async. 

> also `wait_for` returning `Ref<...>` is kind of a pain

I see, would require cloning in the `get` impl to not have `Ref` in the public interface :/

Reply:
> I think that's a bad idea since it mixes sync+async.

Right. It will have to a different method.